### PR TITLE
Factor code related to class hint clenv

### PR DESCRIPTION
**Pull request by ppedrot**
This PR factorizes quite a bit of code around hint handling, specifically in typeclasses eauto. An expensive evarmap refreshing call has been removed in the process, and the internal hint type was made private.

Overlays:
- https://github.com/lukaszcz/coqhammer/pull/76

### CODE REVIEW

**Comment by ppedrot** [comment]
I ran a bench just in case:
```
┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬────────────────────────────────┬───────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]      │    mem faults     │
│                             │                         │                                             │                                             │                                │                   │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD  PDIFF   │  NEW  OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-bedrock2 │  224.14  225.72 -0.70 % │      620630032236      624223789986 -0.58 % │     1032654835267     1032767549965 -0.01 % │    1092004    1091940  +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-character │  163.36  164.37 -0.61 % │      454940636201      456868108957 -0.42 % │      633682570665      633653203598 +0.00 % │     730212     728160  +0.28 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  142.79  143.41 -0.43 % │      397574755482      398598190395 -0.26 % │      517446357229      517407415650 +0.01 % │     562844     562932  -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-odd-order │ 1131.80 1136.47 -0.41 % │     3150834536072     3165216818107 -0.45 % │     5093728341499     5093302452956 +0.01 % │    1031616    1120796  -7.96 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-coqprime │   83.71   84.04 -0.39 % │      229878297060      230556366468 -0.29 % │      324922087987      324799622699 +0.04 % │     769960     769668  +0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                         coq │  564.93  566.97 -0.36 % │     1583817369776     1588824638301 -0.32 % │     2160176023244     2160051159182 +0.01 % │     604644     604668  -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-fiat-parsers │  670.42  672.44 -0.30 % │     1878294849982     1882274078141 -0.21 % │     2835152644976     2836370836092 -0.04 % │    3671764    3676708  -0.13 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fourcolor │ 2514.48 2521.14 -0.26 % │     7011544274003     7030466034766 -0.27 % │    11471782836367    11468264417161 +0.03 % │     766148     766316  -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   49.56   49.68 -0.24 % │      137579217935      137825967324 -0.18 % │      190744340490      190734975568 +0.00 % │     486372     486432  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-unimath │ 4050.31 4060.06 -0.24 % │    11269153870867    11291596230383 -0.20 % │    21549257682833    21548544263704 +0.00 % │    1008900    1006424  +0.25 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-coquelicot │   73.41   73.58 -0.23 % │      201482423912      201590641062 -0.05 % │      262895511430      262690098374 +0.08 % │     608324     608388  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-performance-tests │ 2995.95 3001.96 -0.20 % │     8349931012633     8362673190985 -0.15 % │    15178490461513    15173996866520 +0.03 % │    4316532    4316716  -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│ coq-rewriter-perf-SuperFast │ 1034.89 1036.87 -0.19 % │     2878787596548     2883846589515 -0.18 % │     4604887548532     4605412896955 -0.01 % │    1074404    1074240  +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                  coq-geocoq │ 1504.48 1507.27 -0.19 % │     4193929530587     4200910956064 -0.17 % │     6401663093116     6397242889793 +0.07 % │    1216656    1218720  -0.17 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-corn │ 1597.31 1600.18 -0.18 % │     4454307571342     4459357697232 -0.11 % │     6864099600427     6859775747153 +0.06 % │     843928     843956  -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-verdi-raft │ 1272.97 1275.20 -0.17 % │     3546382276866     3549997852621 -0.10 % │     4979662582293     4978725458088 +0.02 % │     976712     976832  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-engine-bench │ 6364.68 6372.05 -0.12 % │    17768142165254    17790573880327 -0.13 % │    31396731501445    31416494117260 -0.06 % │    4389380    4389356  +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-verdi │  121.56  121.70 -0.12 % │      336250463892      336835864962 -0.17 % │      421384321205      421103238568 +0.07 % │     564636     564636  +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│          coq-mathcomp-field │  264.21  264.51 -0.11 % │      734878260291      736322013454 -0.20 % │     1123662301611     1123670795885 -0.00 % │     727552     727336  +0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-bignums │   59.24   59.30 -0.10 % │      163398272743      164127412215 -0.44 % │      221154449329      221111235694 +0.02 % │     474584     474936  -0.07 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-color │  684.13  684.48 -0.05 % │     1907698639880     1910454135589 -0.14 % │     2457097495364     2454040189790 +0.12 % │    1489764    1501268  -0.77 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  175.99  176.08 -0.05 % │      489427605417      490423989776 -0.20 % │      693107880281      692915648075 +0.03 % │     685620     700464  -2.12 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│             coq-lambda-rust │ 1535.30 1535.58 -0.02 % │     4275433944346     4275619599139 -0.00 % │     6126874084679     6104004410636 +0.37 % │    1216016    1103608 +10.19 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   52.80   52.78 +0.04 % │      144428981262      144320680055 +0.08 % │      186873265651      186862190652 +0.01 % │     522728     517256  +1.06 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-compcert │  695.58  695.22 +0.05 % │     1937453674402     1938623323814 -0.06 % │     2704895929161     2704071881407 +0.03 % │    1139068    1139100  -0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-flocq │  171.41  171.31 +0.06 % │      475765653920      476047904561 -0.06 % │      600771552981      600410431975 +0.06 % │     904580     904416  +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-hott │  323.12  322.84 +0.09 % │      875717175814      876946613225 -0.14 % │     1399287636406     1399374413357 -0.01 % │     619824     621588  -0.28 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                  coq-sf-plf │   44.34   44.30 +0.09 % │      123985084068      124189753621 -0.16 % │      170492655684      170490617852 +0.00 % │     515536     508188  +1.45 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fiat-core │  111.41  111.24 +0.15 % │      317254451933      317654559203 -0.13 % │      425736535686      425575853286 +0.04 % │     494596     494104  +0.10 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-math-classes │  213.51  212.98 +0.25 % │      596727023457      595948260217 +0.13 % │      805003588919      802965811060 +0.25 % │     514688     513428  +0.25 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-rewriter │  584.21  582.75 +0.25 % │     1620140853863     1615563455160 +0.28 % │     2497163582194     2495814082923 +0.05 % │    1041340     995400  +4.62 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-coqutil │   61.84   61.45 +0.63 % │      171005940715      170487389523 +0.30 % │      214932487101      214944257804 -0.01 % │     535000     536100  -0.21 % │    0    0  +nan % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴────────────────────────────────┴───────────────────┘
```

File: tactics/hints.ml [hunk 1]
```diff
@@ -1593,3 +1593,46 @@ struct
...
+    let emap c = Vars.subst_univs_level_constr subst c in
+    let evd = Evd.merge_context_set Evd.univ_flexible evd ctx in
+    (* Only metas are mentioning the old universes. *)
+    let clenv = {
+      templval = Evd.map_fl emap clenv.templval;
+      templtyp = Evd.map_fl emap clenv.templtyp;
+      evd = Evd.map_metas emap evd;
+      env = Proofview.Goal.env gl;
+    } in
+    clenv
```
Comment (SkySkimmer) [inline tactics/hints.ml:1617]:
```suggestion
    {
      templval = Evd.map_fl emap clenv.templval;
      templtyp = Evd.map_fl emap clenv.templtyp;
      evd = Evd.map_metas emap evd;
      env = Proofview.Goal.env gl;
    }
```

File: tactics/class_tactics.ml [hunk 2]
```diff
@@ -144,61 +144,53 @@ let auto_unif_flags ?(allowed_evars = AllowAll) st =
...
-
-let unify_resolve_refine flags gl clenv =
+    let sigma, term = Hints.fresh_hint env sigma h in
+    let ty = Retyping.get_type_of env sigma term in
+    let sigma, cl = Clenv.make_evar_clause env sigma ?len ty in
+    let term = applist (term, List.map (fun x -> x.hole_evar) cl.cl_holes) in
+    let sigma =
+      Evarconv.(unify_leq_delay
+                  ~flags:(default_flags_of flags.core_unify_flags.modulo_delta)
+                            env sigma cl.cl_concl concl)
```
Comment (SkySkimmer) [inline tactics/class_tactics.ml:185]: This looks weirdly indented, shouldn't `env` be at the same indent as `~flags`?

**Review by SkySkimmer — APPROVED** [review]
Looks good for universes.

File: tactics/class_tactics.ml (hunk 2 above)
Comment (ppedrot) [reply tactics/class_tactics.ml:185]: This comes from the original piece of code.

**Comment by ppedrot** [comment]
This needs an assignee.

**Review by herbelin — APPROVED** [review]
This is part of a sequence of cleaning and efficiency improvements which I expect to be positive. I'm not a specialist of class hints but I'm ready to approve to allow the process to continue. I'll merge tomorrow if noone has more to say by then.

@ppedrot: what about integrating @skyskimmer's short code simplification?

**Comment by ppedrot** [comment]
@herbelin will do.

**Comment by ppedrot** [comment]
This is now ready.

**Pull request merged**